manta-benchmark/benches/ec.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+use ark_bls12_381::{g1::Parameters, G1Affine, G1Projective};
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use manta_benchmark::ec;
+use manta_crypto::rand::OsRng;
+
+fn affine_affine_addition(c: &mut Criterion) {
+    let mut group = c.benchmark_group("bench");
+    let mut rng = OsRng;
+    let mut lhs = black_box(ec::sample_affine_point::<G1Affine, _>(&mut rng));
+    let rhs = black_box(ec::sample_affine_point(&mut rng));
+    group.bench_function("affine-affine addition", |b| {
+        b.iter(|| {
+            ec::affine_affine_add_assign(&mut lhs, &rhs);
+        })
+    });
+}
+
+fn projective_affine_addition(c: &mut Criterion) {
+    let mut group = c.benchmark_group("bench");
+    let mut rng = OsRng;
+    let mut lhs = black_box(ec::sample_projective_point::<G1Projective, _>(&mut rng));
+    let rhs = black_box(ec::sample_affine_point(&mut rng));
+    group.bench_function("projective-affine addition", |b| {
+        b.iter(|| {
+            ec::projective_affine_add_assign(&mut lhs, &rhs);
+        })
+    });
+}
+
+fn projective_projective_addition(c: &mut Criterion) {
+    let mut group = c.benchmark_group("bench");
+    let mut rng = OsRng;
+    let mut lhs = black_box(ec::sample_projective_point::<G1Projective, _>(&mut rng));
+    let rhs = black_box(ec::sample_projective_point::<G1Projective, _>(&mut rng));
+    group.bench_function("projective-projective addition", |b| {
+        b.iter(|| {
+            ec::projective_projective_add_assign(&mut lhs, &rhs);
+        })
+    });
+}
+
+fn affine_scalar_multiplication(c: &mut Criterion) {
+    let mut group = c.benchmark_group("bench");
+    let mut rng = OsRng;
+    let point = black_box(ec::sample_affine_point::<G1Affine, _>(&mut rng));
+    let scalar = black_box(ec::sample_scalar::<Parameters, _>(&mut rng));
+    group.bench_function("affine-scalar multiplication", |b| {
+        b.iter(|| {
+            let _ = ec::affine_scalar_mul(&point, scalar);
```

Comment:
Not necessary though, but it might be better to add black-box to result as well?
```suggestion
            let result = ec::affine_scalar_mul(&point, scalar);
            black_box(result)
```

Reply:
This is probably fine since we have already added `blackbox` on inputs `point` and `scalar`. The profiling results also seem to make sense.

I am conservative on adding `blackbox` in the `b.iter(...)` since [`it may cause some performance overhead`](https://docs.rs/criterion/0.2.1/criterion/fn.black_box.html).

Reply:
Just for your interest, here is an experiment I just did: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=bc815c295fbcbe58a5ac8b05d8fadf60

Try to compile the code in release mode and see the x86 assembly of `test1` and `test2`, you can see that without black-box on the result, the entire computation just goes away.